[BEAM-5633] Adds reconnect logic to python logging client

mwylde opened the pull request:
This PR attempts to fix BEAM-5633 (Python SDK harness logging client failure) by adding reconnect logic to the logging client. With this patch, we haven't seen logging failures on our infrastructure.

This issue manifests itself on the server (taskmanager) side with the exception

```
[grpc-default-executor-0] WARN org.apache.beam.runners.fnexecution.logging.GrpcLoggingService - Logging client failed unexpectedly.
org.apache.beam.vendor.grpc.v1.io.grpc.StatusRuntimeException: CANCELLED: cancelled before receiving half close
	at org.apache.beam.vendor.grpc.v1.io.grpc.Status.asRuntimeException(Status.java:517)
	at org.apache.beam.vendor.grpc.v1.io.grpc.stub.ServerCalls$StreamingServerCallHandler$StreamingServerCallListener.onCancel(ServerCalls.java:272)
```

and on the client (python worker) side with

```
E1004 22:08:02.139750634      38 chttp2_transport.cc:1097]   Received a GOAWAY with error code ENHANCE_YOUR_CALM and debug data equal to "too_many_pings"
Exception in thread read_log_control_messages:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/local/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/site-packages/apache_beam/runners/worker/log_handler.py", line 65, in <lambda>
    target=lambda: self._read_log_control_messages(log_control_messages),
  File "/usr/local/lib/python2.7/site-packages/apache_beam/runners/worker/log_handler.py", line 111, in _read_log_control_messages
    for _ in log_control_iterator:
  File "/usr/local/lib/python2.7/site-packages/grpc/_channel.py", line 350, in next
    return self._next()
  File "/usr/local/lib/python2.7/site-packages/grpc/_channel.py", line 341, in _next
    raise self
_Rendezvous: <_Rendezvous of RPC that terminated with (StatusCode.UNAVAILABLE, Socket closed)>
```

At a high-level, the issue is the client is sending the server too many illegal pings, and the server eventually closes the connection. As the client doesn't have any retry logic it never recovers from this.

Ideally, there'd be two levels of fixes:

1. Update the server or client gRPC configuration to prevent the client from sending bad pings or to prevent the server from killing the connection in this situation (see [here](https://github.com/grpc/grpc/blob/master/doc/keepalive.md) for more documentation of this behavior)
2. Make the logging client more robust, reconnecting when the connection is lost

I haven't been able to find a configuration that solves the first part, so that may need more investigation. Regardless, I think it's worthwhile having reconnection logic even if that issue is fixed.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

aaltay commented on sdks/python/apache_beam/runners/worker/log_handler.py:
```diff
@@ -108,5 +113,16 @@ def _write_log_entries(self):
 
   def _read_log_control_messages(self, log_control_iterator):
```
How about removing `log_control_iterator` as an argument? (Since `_read_log_control_messages` can call connect directly.)

aaltay commented on sdks/python/apache_beam/runners/worker/log_handler.py:
```diff
@@ -108,5 +113,16 @@ def _write_log_entries(self):
+        return
+      except Exception as ex:
+        # reset the stream
+        print >> sys.stderr, "Logging client failed: {}... resetting".format(ex)
```
This `print` is not python 3 compatible. Please use the print(...) syntax.

rakeshcusat replied:
+1^

rakeshcusat replied:
```
print("Logging client failed: {}... resetting".format(ex), file=sys.stderr)
```
If you want to print the stack trace then syntax could be little different.

mwylde replied:
Fixed

aaltay commented on sdks/python/apache_beam/runners/worker/log_handler.py:
```diff
@@ -108,5 +113,16 @@ def _write_log_entries(self):
+      except Exception as ex:
+        # reset the stream
+        print >> sys.stderr, "Logging client failed: {}... resetting".format(ex)
+        self.acquire()
```
is this to prevent emit from running while this is happening?

mwylde replied:
Yeah, that was the idea. However digging a bit deeper I think this is undesirable, as it will block application logic during the reconnection process. Given that we have a queue in between the emit call and the grpc stream, it shouldn't be necessary to lock.

mwylde replied:
However, currently the queue is unbounded, so if we remove this we'd need to bound the queue and either block when it's full or drop messages. Given that logging is a secondary concern, I think it may make more sense to drop the messages. Thoughts?

aaltay replied:
It sounds reasonable to me to put an upper bound on logs that might be buffered when disconnected. 
- The upper bound of 1000 messages is arbitrary (any number will be arbitrary though). If we take your estimate of 1K/log message, we can probably increase that a bit.
- It would help to have a log message after reconnecting saying that some message might have been dropped in between. Most people reading a set of logs will assume its completeness.

mwylde replied:
Yeah, happy to change the buffer size according to what level of memory usage people find reasonable. 10,000?

I like the idea of logging a message on reconnection. Ideally we'd also have a metric for tracking dropped logs, but my understanding is that there isn't currently metric support in the python SDK.

mwylde commented:
I've updated with the review comments and fixed lint failures. There's an open question about what the behavior should be when the logging connection is down (e.g., when we're in the process of reconnecting). We can queue up some number of logs, but past that point should we block execution or drop logs?

Currently this PR queues up to 1,000 logs and drops past that point. The queue size is semi-arbitrary, but to me it makes sense to prioritize pipeline execution over capturing 100% of logs. Would be interested in hearing other opinions on this, though.

tweise commented:
@aaltay @robertwb PTAL

aaltay commented on sdks/python/apache_beam/runners/worker/log_handler.py:
```diff
@@ -77,14 +83,14 @@ def emit(self, record):
     # released.
     self.acquire()
-    self._log_entry_queue.put(self._FINISHED)
+    self._log_entry_queue.put(self._FINISHED, timeout=1)
```
I think we need this to be blocking, to ensure that existing buffers is fully flushed out.

mwylde replied:
The default is blocking, so we'll block for up to 1 second. The issue I see is that we probably don't want to indefinitely block shutdown in the case where we're unable to re-connect to the logging service.

aaltay replied:
Sounds reasonable.

aaltay commented on sdks/python/apache_beam/runners/worker/log_handler.py:
```diff
@@ -106,7 +120,19 @@ def _write_log_entries(self):
+    while True:
+      log_control_iterator = self.connect()
+      if self._dropped_logs > 0:
+        logging.warn("Dropped %d logs while logging client disconnected",
```
I do not want to block this PR. Let's merge. but I have a follow up question.

If the queue is already full, would not this message be also dropped as well. (Unless the log write gets scheduled in between connection and this log message?

mwylde replied:
It's not guaranteed to make it, although in my tests it consistently does. Once we call connect() we start to drain from the queue. So so long as something has drained by the point we get to the warn it will show up.

Alternatively, we could write to stderr as we do to report the connection error as that should be reliable (but might not be as accessible as real logs). Happy to hear other suggestions as well.

aaltay commented:
Thank you @mwylde. I will merge. I left you a follow up question though.

mwylde commented:
Thanks for the review!

angoenka commented:
We fixed keep alive issue by explicitly setting the keep alive time on server and clients in pr https://github.com/apache/beam/pull/7415

I am planning to revisit the reconnect logic as I see that the connect method is called multiple times on pipeline termination causing hang of the sdk harness container.
